### ISSUE TITLE
Some context is lost with a combination of extends, embed, include

### ISSUE THREAD

**Issue by tremby** [twigjs/twig.js#287]
I am losing some data variables in a certain situation on my site.

I gradually boiled the case down and have posted it in a gist: https://gist.github.com/tremby/62df667cf9fb9dc7a303

Pasting the output of the `test2page.twig` template here:

```
test2layout.twig 1      [foo]   [bar]
test2page.twig 1        [foo]   []
test2.twig 1            [foo]   []      [baz]
test2template.twig 1    []      []      [baz]
test2.twig 2            []      []      [baz]
test2template.twig 2    []      []      [baz]
test2.twig 3            [foo]   []      [baz]
test2page.twig 2        [foo]   []
test2layout.twig 2      [foo]   [bar]
```

The variable foo is set to 'foo' as part of the render function and the way I understand it, it should be available everywhere (I'm not doing include with "only" at any point). But it's not available in the block `test2.twig` embeds within `test2template.twig`, or in the non-block parts of `test2template.twig` embedded in `test2.twig`. (This is the main problem.)

bar is set to 'bar' in the layout which `test2page.twig` extends, and I'm not sure whether that should then be available in `test2page.twig` and the file that this one includes. I would have thought it should, but it is not.

baz is set to `baz` in `test2.twig`, which is included by `test2page.twig`. It's then available everywhere I'd expect.

Is this indeed a bug, or am I misunderstanding how things are supposed to work?

**Comment by evgenius** [comment]
@tremby There may be a bug, thank you for posting this issue.
The way the context is handled is different compared to Twig.php so the more cases and tests we have, the closer its behavior will be.

**Comment by dotmagic** [comment]
14 months later i run into the same issue :(
My workaround was to use `include` and pass data with `with` instead `embed`.
Actually the embed has a very limited use.

**Comment by smutnyleszek** [comment]
Same problem here. It basically boiled down to this:

```twig
{% set foo = '123' %}

{% embed 'test.html.twig' with {
    bar: '456'
} %}
    {% block attributes %}
        {{ foo }} {# <-- here foo is undefined #}
    {% endblock attributes %}
{% endembed %}
```

My fix was to pass `foo` with `with` object

**Comment by olets** [comment]
@smutnyleszek what version of twig.js are you using, and how are you using it?

**Comment by smutnyleszek** [comment]
@olets twig version is `1.10.5` and I run it via `grunt` with this options:

```
const template = Twig.twig({
    strict_variables: true,
    id: templateName,
    base: VIEWS_PATH,
    namespaces: {
        nodemodules: 'node_modules/'
    },
    path: twigPath,
    rethrow: true,
    async: false
});
html = template.render({
    foo: FOO_ENV,
});
grunt.file.write(PATH, html);
```

**Comment by olets** [comment]
@smutnyleszek rats I can't reproduce this (1.10.5 via `gulp-twig`). I used this generic version of your situation

```twig
{# main.twig #}
{% set x = 1 %}
{% set y = 2 %}

{% embed 'embedded.twig' %}
    {% block test %}
        x: {{ x }}
    {% endblock %}
{% endembed %}
```

```twig
{# embedded.twig #}
{% block test %}
   y: {{ y }}
{% endblock %}
```

My output: `x: 1`, identical to Twig.php's (see <https://twigfiddle.com/sxrdwt>)

**Comment by smutnyleszek** [comment]
Ok, I probably simplified the example too much. 

I took the real code this time and kept all the things from the embed/include chain:

```twig
{# Settings/settings.html.twig #}
{% spaceless %}

<!doctype html>
<html lang="en-US">

<head>
    <meta charset="UTF-8">
    <title>&nbsp;</title>
    {% block additionalStyles %}{% endblock %}
</head>

<body>
{% include 'Settings/_settingsContent.html.twig' %}
</body>

</html>
{% endspaceless %}
```

```twig
{# Settings/_settingsContent.html.twig #}
<section>
    {% include 'Settings/sections/_general.html.twig' %}
</section>

```twig
{# Settings/sections/_general.html.twig #}
{% set applicationType = 'GogGalaxy' %}
{% if applicationType == 'GogGalaxy' %}
    {# ↓↓↓ SOLUTION: pass isForbiddenToUndoPublicBeta as embed param ↓↓↓ #}
    {% embed 'Settings/partials/_settingRow.html.twig' with {
        label: 'settings__label__public_beta',
        isForbiddenToUndoPublicBeta: isForbiddenToUndoPublicBeta
    } %}
        {% block settingRowContent %}
            {# ↓↓↓ SOLUTION: pass isForbiddenToUndoPublicBeta as embed param ↓↓↓ #}
            {% embed 'Settings/partials/_settingCheckbox.html.twig' with {
                settingName: 'publicBetaEnabled',
                label: 'settings__text__public_beta',
                isForbiddenToUndoPublicBeta: isForbiddenToUndoPublicBeta
            } %}
                {% block labelAttributes %}
                    {# ↓↓↓ PROBLEM: isForbiddenToUndoPublicBeta was undefined ↓↓↓ #}
                    {% if isForbiddenToUndoPublicBeta %}
                        ng-disabled="clientSettings.settingsData.publicBetaEnabled"
                    {% endif %}
                {% endblock labelAttributes %}
            {% endembed %}

            {# ↓↓↓ PROBLEM: isForbiddenToUndoPublicBeta was undefined ↓↓↓ #}
            {% if isForbiddenToUndoPublicBeta %}
                <p
                    i-settingRow-description
                    ng-cloak
                    ng-show="clientSettings.settingsData.publicBetaEnabled"
                    translate="settings__text__public_beta_not_undoable_new"
                    translate-values="{
                        linkStart: '<a i-settingRow-descriptionLink href=\'https://support.gog.com/hc/articles/115002280365\' target=\'_blank\'>',
                        linkEnd: '</a>'
                    }"
                ></p>
            {% endif %}
        {% endblock settingRowContent %}
    {% endembed %}
{% endif %}
```

```twig
{# Settings/partials/_settingRow.html.twig #}
<div {% block settingRowAttributes %}{% endblock %}>
    <div i-settingRow-label translate="{{ label }}"></div>
    <div i-settingRow-content>{% block settingRowContent %}{% endblock %}</div>
</div>
```

```twig
{# Settings/partials/_settingCheckbox.html.twig #}
<label {% block labelAttributes %}{% endblock %}>
    <input type="checkbox">
    <span translate="{{ label }}"></span>
</label>
```

**Comment by olets** [comment]
@smutnyleszek could you reduce that to a minimal generic example that produces the problem? That'll help pinpoint the problem.

**Comment by toptalo** [comment]
Maybe this issue related to issue #352  and my PR #537

**Comment by willrowe** [comment]
There are a bunch of issues with how the context is handled in embeds, blocks, etc. As I mentioned in #534, I have a complete re-implementation of how blocks are handled ready to go, I'm just waiting on some movement in this repo to merge PRs.

**Comment by toptalo** [comment]
Okay.
I will wait

**Comment by dave-irvine** [comment]
I'll try and take a look tonight, I keep forgetting when I get in from work

**Comment by toptalo** [comment]
@dave-irvine Hello! Any news?

**Comment by dave-irvine** [comment]
Hi,

Sorry I haven't looked. I will set a reminder in my phone to look tonight.

**Comment by toptalo** [comment]
Thanks! I really need it to use embeds in templates

**Comment by mh-nichts** [comment]
Hello, I apparently have the same issue, so I would be happy to see it resolved as well.

**Comment by olets** [comment]
🎉

**Comment by toptalo** [comment]
Still have issue with embeds

**Comment by willrowe** [comment]
I'm working on some other fixes to embeds.
Which issue are you still encountering? I want to make sure I'm covering the correct cases.

**Comment by toptalo** [comment]
I have page template that extends base template. And if I include embed inside page template I lose content of blocks that are defined upper.
My PR #537 resolve my problem, but it can have some side effects.

**Comment by willrowe** [comment]
Ok, I think I have a more robust fix for that.
Please provide a TwigFiddle link of a basic test case that is not working for you so I can confirm it fixes your issue.

**Comment by toptalo** [comment]
Here it is https://twigfiddle.com/960szz

Here is result that I get in my env

```html
<!doctype html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Document</title>
</head>
<body>
    <header>
        <h1>Title</h1>
    </header>
    <main>
        <article>
            <header>
            </header>
            <div class="content">
                <section class="section section_article">
                    <h3>Section Title</h3>
                    <div class="section_content">
                        <p>Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.</p>
                    </div>
                </section>
            </div>
            <footer>
                <p>Author: John Doe</p>
            </footer>
        </article>
    </main>
    <footer>
        <small>&copy; 2018</small>
    </footer>
</body>
</html>
``` 

`article > header` is empty

**Comment by willrowe** [comment]
Thanks for doing that, I'll include this when I'm testing.

**Comment by mh-nichts** [comment]
Hello, I just want to confirm that the problem is also solved for me when using v1.12.0.
Thanks !

**Comment by willrowe** [comment]
This was fixed by #550.
